lib/compact.js

```diff
@@ -308,40 +313,100 @@ api.compact = ({
+            // wrap using @graph alias
+            compactedItem = {
+              [api.compactIri({activeCtx, iri: '@graph'})]: compactedItem
+            };
+
+            // include @id from expanded graph, if any
+            if('@id' in expandedItem) {
+              compactedItem[api.compactIri({activeCtx, iri: '@id'})] =
+                expandedItem['@id'];
+            }
 
-        // handle language and index maps
-        if(container.includes('@language') || container.includes('@index')) {
+            // include @index from expanded graph, if any
+            if('@index' in expandedItem) {
+              compactedItem[api.compactIri({activeCtx, iri: '@index'})] =
+                expandedItem['@index'];
+            }
+            _addValue(
+              rval, itemActiveProperty, compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          }
+        } else if(container.includes('@language') || container.includes('@index') ||
+          container.includes('@id') || container.includes('@type')) {
+          // handle language and index maps
           // get or create the map object
-          let mapObject;
+          let mapObject, key;
           if(itemActiveProperty in rval) {
             mapObject = rval[itemActiveProperty];
           } else {
             rval[itemActiveProperty] = mapObject = {};
           }
 
+          if(container.includes('@language')) {
           // if container is a language map, simplify compacted value to
           // a simple string
-          if(container.includes('@language') && _isValue(compactedItem)) {
-            compactedItem = compactedItem['@value'];
+            if(_isValue(compactedItem)) {
+              compactedItem = compactedItem['@value'];
+            }
+            key = expandedItem['@language'];
+          } else if(container.includes('@index')) {
+            key = expandedItem['@index'];
+          } else if(container.includes('@id')) {
+            const idKey = api.compactIri({activeCtx, iri: '@id', vocab: true});
+            key = compactedItem[idKey] || '@none';
+            delete compactedItem[idKey];
+          } else if(container.includes('@type')) {
+            const typeKey = api.compactIri({activeCtx, iri: '@type', vocab: true});
+            var types;
+            [key, ...types] = [].concat(compactedItem[typeKey] || ['@none']);
+            switch(types.length) {
+            case 0: delete compactedItem[typeKey]; break;
+            case 1: compactedItem[typeKey] = types[0]; break;
+            default: compactedItem[typeKey] = types; break;
+            }
           }
 
           // add compact value to map object using key from expanded value
           // based on the container type
-          const c = container.includes('@language') ? '@language' : '@index';
-          _addValue(mapObject, expandedItem[c], compactedItem);
+          _addValue(mapObject, key, compactedItem,
```

Comment:
Our coding style rule for long parameter lists is:

When moving any function parameters down a line, move them all down a line (and then add new lines as needed to fit within 80-chars, but do not leave any on the same line as the function as it improves readability):

```js
_addValue(
  mapObject, key, compactedItem, {propertyIsArray: container.includes('@set')});
```